### PR TITLE
chore: break down WorkloadManager [DET-5419]

### PR DESCRIPTION
**Pull request by rb-determined-ai**
## Description

Remove the determined.callback.Callback, which only had one
implementation in existence.  Just inline those calls.

Reduce the WorkloadManager to a trio of hooks, and attach them to a new
_GenericContext.  Eventually _GenericContext will be exposed as a public
API and will become the basis of the Generic API.

Each TrialController instantiates its own _GenericContext, which is
basically how the TrialControllers will work in the future.

Significantly, this moves all major logic out of the parent processes
and into the the worker processes.  All that remains in the parent
processes is the SocketManager and the calling of horovodrun.  The
SocketManager will be removed after push architecture is complete,
leaving only horovodrun, which will eventually evolve into a
configurable launch layer.

## Test Plan

Mostly existing tests should still pass.  Since the code heavily affects the dtrain architecture, running parallel tests is necessary before landing the PR.

## Commentary (optional)

If you replace each part of a ship, one-by-one, is it still the same ship?

It's hard to tell.

Similarly, it's hard to tell if the `WorkloadManager` disappeared or if it was renamed `_GenericContext`

### CODE REVIEW

**Comment by rb-determined-ai** [comment]
@shiyuann I think this is ready for review. Let me know if you see parts in there that you would like me to pull out into their own PR.

Also, this PR does not break down the EnvContext or ExperimentConfig, which are basically still passed around everywhere.  That's still a problem but I don't think now is the right time to deal with that problem, and I don't think this PR makes that problem any worse.

I do think it is a problem we need to address before making the Generic API public, and I put some thought into how we would break those objects up into smaller components, where different parts of the system accessed only their relevant components.  See [here](https://gist.github.com/rb-determined-ai/80f5ce24ad0178cb676eccef4d8aa34e) for my thoughts on that.

File: harness/tests/test_ipc.py [hunk 1]
```diff
@@ -120,20 +120,22 @@ def health_check() -> None:
 
 
 def test_subprocess_launcher_receiver() -> None:
-    env = utils.make_default_env_context(hparams={"global_batch_size": 1})
+    utils.make_default_exp_config
```
Comment (shiyuann) [inline harness/tests/test_ipc.py:123]:
```suggestion
```

File: harness/determined/experimental/_native.py [hunk 2]
```diff
@@ -89,29 +89,27 @@ def _submit_experiment(
...
-    yield from interceptor.send(workload.validation_workload(1), [])
+    yield from interceptor.send(workload.validation_workload(1))
     validation = interceptor.metrics_result()
     v_metrics = validation["metrics"]["validation_metrics"]
     logging.info(f"Finished validating, validation metrics: {v_metrics}")
 
-    logging.info(f"Saving a checkpoint to {checkpoint_dir}.")
-    yield workload.checkpoint_workload(), [checkpoint_dir], workload.ignore_workload_response
-    logging.info(f"Finished saving a checkpoint to {checkpoint_dir}.")
+    logging.info("Saving a checkpoint to.")
```
Comment (shiyuann) [inline harness/determined/experimental/_native.py:108]: This should be fixed

File: harness/tests/experiment/fixtures/tf_keras_one_var_model.py [hunk 3]
```diff
@@ -22,6 +22,7 @@ def make_one_var_tf_dataset_loader(hparams: Dict[str, Any], batch_size: int):
 
 
 class OneVarTrial(keras.TFKerasTrial):
+    _searcher_metric = "loss"
```
Comment (shiyuann) [inline harness/tests/experiment/fixtures/tf_keras_one_var_model.py:25]: I didn't see those `_searcher_metric` fields being used.

File: harness/determined/_experiment_config.py [hunk 4]
```diff
@@ -46,3 +46,10 @@ def get_min_validation_period(self) -> Dict:
         min_validation_period = self.get("min_validation_period", {})
         assert isinstance(min_validation_period, dict)
         return min_validation_period
+
+    def get_searcher_metric(self) -> str:
+        searcher_metric = self.get("searcher", {}).get("metric")
+        assert isinstance(
```
[the remark this answers is not in the document]
Comment (rb-determined-ai) [reply harness/determined/_experiment_config.py:52]: Nothing.  The assert is mostly for mypy.  But the error message was weird, so I made it clearer.

File: harness/determined/_train_context.py [hunk 5]
```diff
@@ -312,3 +315,92 @@ def _zmq_allgather(self, stuff: Any) -> List:
...
+        else:
+
+            @functools.wraps(fn)
+            def _fn(*_: Any, **__: Any) -> Any:
+                out = self._zmq_broadcast_local(None)
+                return out
+
+        return _fn
+
+    def _local_chief_contextmanager(self, fn: Callable) -> Callable:
```
[the remark this answers is not in the document]
Comment (rb-determined-ai) [reply harness/determined/_train_context.py:375]: I don't think so, because I wanted to mirror the word `contextmanager` from `contextlib.contextmanager`

File: harness/determined/_train_context.py [hunk 6]
```diff
@@ -312,3 +315,92 @@ def _zmq_allgather(self, stuff: Any) -> List:
...
+
+        Not efficient; built on top of _zmq_allgather().
+        """
+        machine_rank = self._rendezvous_info.get_rank()
+        all_stuff = self._zmq_allgather((machine_rank, self.get_local_rank(), stuff))
+        local_stuff = [s for s in all_stuff if s[0] == machine_rank and s[1] == 0]
+        assert len(local_stuff) == 1, f"invalid allgather machine_rank={machine_rank}: {all_stuff}"
+        return local_stuff[0]
+
+    def _local_chief_fn(self, fn: Callable) -> Callable:
```
[the remark this answers is not in the document]
Comment (rb-determined-ai) [reply harness/determined/_train_context.py:353]: done.  Also removed _zmq_allgather_local, which was unused.

File: harness/determined/_train_context.py [hunk 7]
```diff
@@ -312,3 +315,92 @@ def _zmq_allgather(self, stuff: Any) -> List:
...
+        """
+        if not self._hvd_config.use:
+            return stuff
+        if self._is_chief:
+            self._chief_zmq.broadcast(stuff)
+        else:
+            stuff = self._worker_zmq.recv()
+        return stuff
+
+    def _zmq_broadcast_local(self, stuff: Any = None) -> Any:
```
[the remark this answers is not in the document]
Comment (rb-determined-ai) [reply harness/determined/_train_context.py:341]: you mean to make it efficient?  I actually think we should make it communicate over unix sockets, but that's quite a bit of work for something that only applies to checkpoint coordination right now.  So no, not soon.

File: harness/determined/_generic.py [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+            env.det_experiment_id,
+            env.det_trial_id,
+            env.experiment_config["checkpoint_storage"],
+            container_path=None if not env.on_cluster else constants.SHARED_FS_CONTAINER_PATH,
+        )
+
+        self._tensorboard_writer = tensorboard.get_metric_writer()
+
+    @contextlib.contextmanager
+    def _load_initial_checkpoint(self, checkpoint: Dict) -> Iterator[pathlib.Path]:
```
[the remark this answers is not in the document]
Comment (rb-determined-ai) [reply harness/determined/_generic.py:45]: done.

File: harness/tests/test_ipc.py (hunk 1 above)
Comment (rb-determined-ai) [reply harness/tests/test_ipc.py:123]: lol

File: harness/determined/experimental/_native.py (hunk 2 above)
Comment (rb-determined-ai) [reply harness/determined/experimental/_native.py:108]: fixed.

File: harness/tests/experiment/fixtures/tf_keras_one_var_model.py (hunk 3 above)
Comment (rb-determined-ai) [reply harness/tests/experiment/fixtures/tf_keras_one_var_model.py:25]: Yeah, this was a bit hacky, it's used to automatically construct experiment configs with the proper searcher metric set.  That used to not be important because the workload manager did the searcher metric checks, and that used to not be in the unit tests.

```python
def make_trial_controller_from_trial_implementation(
    trial_class: Type[det.Trial],
    hparams: Dict,
    workloads: workload.Stream,
    scheduling_unit: int = 1,
    trial_seed: int = 0,
    exp_config: Optional[Dict] = None,
    checkpoint_dir: Optional[str] = None,
    latest_checkpoint: Optional[Dict[str, Any]] = None,
) -> det.TrialController:
    if not exp_config:
        assert hasattr(
            trial_class, "_searcher_metric"
        ), "Trial classes for unit tests should be annotated with a _searcher_metric attribute"
        searcher_metric = trial_class._searcher_metric  # type: ignore
        exp_config = make_default_exp_config(
            hparams, scheduling_unit, searcher_metric, checkpoint_dir=checkpoint_dir
        )
```

I thought it was better to annotate the test fixture models with an extra attribute rather than hand-code the searcher metric in the experiment configs everywhere that each model was referenced in tests.

File: harness/determined/_train_context.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (shiyuann) [reply harness/determined/_train_context.py:331]: `_zmq_allgather` is blocking all containers on all machines. All the local chief workers broadcast messages to all local non-chief workers.  Every worker gets a number of the same messages. Are we going to have other functions that use the mechanism?

[the remark this answers is not in the document]
Comment (rb-determined-ai) [reply harness/determined/_train_context.py:331]: not unless we expose something like this in the Generic API.  I'm ok putting off fixing this until then.

I would not want to make it a public API before fixing the underlying mechanism.

[the remark this answers is not in the document]
Comment (rb-determined-ai) [reply harness/determined/_train_context.py:520]: update: this has now been fixed by pulling some ZMQ rework items into their own PR, which this PR is now rebased on top of: https://github.com/determined-ai/determined/pull/2593

File: harness/determined/tensorboard/base.py [hunk 9]
```diff
@@ -61,3 +63,18 @@ def delete(self) -> None:
         Delete all objects from the backing persistent storage.
         """
         pass
+
+
+def get_metric_writer() -> tensorboard.BatchMetricWriter:
```
Comment (shiyuann) [inline harness/determined/tensorboard/base.py:68]: Is it possible to just return an `EventFileWriter` or `SummaryWriter`? The current implementation looks not very clean.

**Review by shiyuann — APPROVED** [review]

**Comment by shiyuann** [comment]
@rb-determined-ai I approved the change on top of #2593 rather than the PR as a whole. So we should try to get #2593 in first.

**Comment by rb-determined-ai** [comment]
I'm closing this PR as it will be appear as part of #2771 and it will land as part of #2727.
